tech-papers/lbft-verification/lbft-verification.tex

```diff
@@ -0,0 +1,199 @@
+The \texttt{ValidatorChangeProof} chain is included in the \texttt{UpdateToLatestLedgerResponse}, hence every participant in the system can maintain the uptodate validator set (given a trusted validator set $VS_i$ one can verify a \texttt{LedgerInfo} with a non-empty \texttt{next\_validator\_set} and thus update its trusted validator set to $VS_{i+1}$). Note that this approach assumes \textbf{some initial} trusted validator set, a challenge thoroughly described in Section~\ref{waypoints}.
+
+The sync up protocol of \texttt{StateSynchronizer} makes sure that both the transactions and the \texttt{ValidatorChangeProof} are properly synchronized:
+\begin{figure}[ht]
+	\centering
+	\includegraphics[width=\textwidth]{figures/state-sync-epochs.pdf}
+	\caption{\footnotesize{Response chunks do not cross epoch boundaries and carry proofs relative to the \texttt{LedgerInfo} of the chunk's epoch.}}
+	\label{fig:state-sync-epochs}
+\end{figure}
+
+\begin{itemize}
+\item \texttt{GetChunkRequest} carries the information about the known version and epoch of the requester.
+\item \texttt{GetChunkResponse} carries chunks that never cross epoch boundaries. In case a chunk terminates the epoch it belongs to, the \texttt{LedgerInfo} in the chunk also includes the \texttt{next\_validator\_set}, which helps the receiver to move to the next epoch.
+\item One interesting aspect worth noting here is that in case a state synchronization request has a predefined target, the chunk response carries either the \texttt{LedgerInfo} of the given target, or the last \texttt{LedgerInfo} of the request epoch (the smaller between the two), to make sure that the receiver can still verify the response as depicted in Figure~\ref{fig:state-sync-epochs}. In these cases the chunk can be smaller than the max allowed chunk size, just to make sure that the chunks do not cross epoch boundaries.
+\item Each intermediate end-of-epoch \texttt{LedgerInfo} is added to the storage after \texttt{StateSynchronizer} verifies that its local accumulator state matches the metadata of this \texttt{LedgerInfo} (this is possible because chunks are executed sequentially and never cross epoch boundaries).
+\end{itemize}
+
+\section{Waypoints}
+\label{waypoints}
+\subsection{Motivation}
+\label{waypoint-motivation}
+A \emph{cold} node (the one that comes online for the first time, or after a long period of being disconnected), faces a problem of the initial syncup because it might no longer trust the old public keys of the historical validators it is aware of, which could have been compromised since then (the so called “long range attack”).
+
+We are using \emph{Waypoints} as an off-chain mechanism for providing the initial trusted set of validators and a point in the accumulator history. Waypoints are going to be published by the association on a regular basis (e.g., on libra.org once in a couple of weeks), and can be embedded within the nodes (clients / full nodes / validators) as a trusted “starting point”.
```

Comment:
Some of this will need to be agreed on by the TSC.  It might be better to omit details until there is agreement.

```suggestion
\emph{Waypoints} can be used as an off-chain mechanism for selecting the initial trusted set of validators and a specific points of blockchain time in the accumulator history. Waypoints can be published by anyone.  The Libra Association is also expected to publish waypoints  on a regular basis (e.g., on libra.org).  Any waypoint selected by clients can be embedded within the nodes (clients / full nodes / validators) as a trusted “starting point”.
```